components/match2/wikis/apexlegends/match_group_input_custom.lua

```diff
@@ -0,0 +1,420 @@
+
+-- containers for process helper functions
+local MatchFunctions = {}
+local MapFunctions = {}
+local OpponentFunctions = {}
+
+local CustomMatchGroupInput = {}
+
+-- called from Module:MatchGroup
+function CustomMatchGroupInput.processMatch(match, options)
+	match = MatchFunctions.parseSetting(match)
+	-- Adjust map data, especially set participants data
+	match = MatchFunctions.adjustMapData(match)
+	match = MatchFunctions.getScoreFromMaps(match)
+
+	-- process match
+	Table.mergeInto(match, MatchFunctions.readDate(match))
+	match = MatchFunctions.getOpponents(match)
+	match = MatchFunctions.getTournamentVars(match)
+	match = MatchFunctions.getVodStuff(match)
+	match = MatchFunctions.getExtraData(match)
+
+	return match
+end
+
+function MatchFunctions.adjustMapData(match)
+	local opponents = Array.mapIndexes(function(idx) return match['opponent' .. idx] end)
+	local mapIndex = 1
+	while match['map' .. mapIndex] do
+		local map = match['map' .. mapIndex]
+		local scores
+		Table.mergeInto(map, MatchFunctions.readDate(map))
+		map = MapFunctions.getParticipants(map, opponents)
+		map = MapFunctions.getOpponentStats(map, opponents, mapIndex)
+		map, scores = MapFunctions.getScoresAndWinner(map, match.scoreSettings)
+		map = MapFunctions.getTournamentVars(map)
+		map = MapFunctions.getExtraData(map, scores)
+
+		match['map' .. mapIndex] = map
+		mapIndex = mapIndex + 1
+	end
+
+	return match
+end
+
+-- called from Module:Match/Subobjects
+CustomMatchGroupInput.processMap = FnUtil.identity
+
+-- called from Module:Match/Subobjects
+CustomMatchGroupInput.processPlayer = FnUtil.identity
+
+--
+--
+function CustomMatchGroupInput.processOpponent(record, timestamp)
+	local opponent = Opponent.readOpponentArgs(record)
+		or Opponent.blank()
+
+	local teamTemplateDate = timestamp
+	-- If date is epoch, resolve using tournament dates instead
+	-- Epoch indicates that the match is missing a date
+	-- In order to get correct child team template, we will use an approximately date and not 1970-01-01
+	if teamTemplateDate == DateExt.epochZero then
+		teamTemplateDate = Variables.varDefaultMulti('tournament_enddate', 'tournament_startdate', NOW)
+	end
+
+	Opponent.resolve(opponent, teamTemplateDate)
+	MatchGroupInput.mergeRecordWithOpponent(record, opponent)
+end
+
+function CustomMatchGroupInput.getResultTypeAndWinner(data, indexedScores)
+	if Table.includes(NP_STATUSES, data.finished) then
+		-- Map or Match wasn't played, set not played
+		data.resulttype = 'np'
+		data.finished = true
+	elseif Logic.readBool(data.finished) then
+		-- Map or Match is marked as finished.
+		-- Calculate and set winner, resulttype, placements and walkover (if applicable for the outcome)
+		local winner
+		indexedScores, winner = CustomMatchGroupInput.setPlacement(indexedScores, data.winner, data.finished)
+		data.winner = data.winner or winner
+	end
+
+	--set it as finished if we have a winner
+	if not Logic.isEmpty(data.winner) then
+		data.finished = true
+	end
```

Comment:
can it happen that there is a winner but the match/map is not finished?
could do `data.finished = Logic.nilOr(Logic.readBoolOrNil(data.winner), Logic.isNotEmpty(data.winner))`

Reply:
For a game, it cannot end without a winner
For a match, not aware of any format where winner != finished

Reply:
okay